### PR TITLE
Add .Net 7 target

### LINKED ISSUE #4194: .NET 7 support
We are currently onboarding .NET 7 (preview) in our project and ran into problems when using `NSwagExe_Net60`:
`System.TypeLoadException: Could not load type 'System.Numerics.INumber1'` (interface added in .NET 7).

The reason for the failure is that NSwag generator runs using .NET 6 runtime and tries use reflection to load .NET 7 dll.

**Is there a plan to add .NET 7 support?**
I guess the PR would be very similar as: https://github.com/RicoSuter/NSwag/pull/3632/

Our hack to test the functionality meanwhile is to change `Microsoft.NETCore.App` and `Microsoft.AspNetCore.App` properties to current preview .NET 7 version (e.g. 7.0.0-rc.2.22472.3) in `tools\Net60\dotnet-nswag.runtimeconfig.json` in the .nuget folder.

### LINKED ISSUE #4219: .Net 7.0.0 now on nuget
Hi - due to the constraint on the .net 7 packages, we can't install NSwag.AspNetCore in a project using .net 7.0.0 packages now.

image

### PR DESCRIPTION
**Pull request by trejjam**
fixes #4219
fixes #4194

### CODE REVIEW

**Comment by sergey-tihon** [comment]
global.json probably should be updated as well
https://github.com/RicoSuter/NSwag/blob/master/global.json#L3

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -32,19 +32,25 @@ jobs:
...
+      - uses: actions/setup-dotnet@v3
         with:
           dotnet-version: 5.0.*
-      - uses: actions/setup-dotnet@v1
+      - uses: actions/setup-dotnet@v3
         with:
           dotnet-version: 6.0.*
+      - uses: actions/setup-dotnet@v3
+        with:
+          dotnet-version: 7.0.*
```
Comment (prom3theu5) [inline .github/workflows/build.yml:52]: You can remove preview annotation now. 7.0.100 sdk is out

**Review by lahma — APPROVED** [review]
Splendid work, looks good to me.

**Comment by RicoSuter** [comment]
Also need to update the npm js file: 
https://github.com/RicoSuter/NSwag/blob/master/src/NSwag.Npm/bin/nswag.js#L11

**Comment by TonyMoonStar** [comment]
Hi,

Some of the 7.0.0 libraries support Net6 and Net7, as example EF Core libraries.
NSwag overlaps on Net6 as
NSwag.AspNetCore -> NSwag.Generation.AspNetCore -> Microsoft.Extensions.DependencyInjection.Abstractions (>= 6.0.0 && < 7.0.0)

Where EF Core dependencies are 7.0.0.

Also with Microsoft.Extensions.Options.

for the Generation should be next may be?
```cs
  <ItemGroup Condition="'$(TargetFramework)' == 'net6.0'">
    <FrameworkReference Include="Microsoft.AspNetCore.App" />
    <PackageReference Include="Microsoft.Extensions.DependencyInjection.Abstractions" Version="[6, 8)" />
    <PackageReference Include="Microsoft.Extensions.Options" Version="[6, 8)" />
  </ItemGroup>
```

---
We tend to use Net6, but since it's a Target 7 updated, could uptake also these changes.

**Comment by trejjam** [comment]
What can possibly go wrong :D
The constraints relaxed

**Review by Mahdi-Joveyni — APPROVED** [review]

**Review by GeorgDangl — APPROVED** [review]

File: src/NSwag.Sample.NET70Minimal/nswag.json [hunk 2]
```diff
@@ -0,0 +1,58 @@
+﻿{
+  "runtime": "Net60",
```
Comment (patrickklaeren) [inline src/NSwag.Sample.NET70Minimal/nswag.json:2]: This seems wrong?

Comment (trejjam) [reply src/NSwag.Sample.NET70Minimal/nswag.json]: Sure, it is. Thanks

**Comment by Dzivo** [comment]
Do you have an idea of release schedule

**Comment by r-work** [comment]
This fails if your project is net 7 and have `<PlatformTarget>x64</PlatformTarget>` in your csproj with the following message:

```
 Detected package downgrade: System.Collections from 4.3.0 to 4.0.11. Reference the package directly from the project to select a different version. 
 MiSystems.GeoServices -> NSwag.AspNetCore 13.17.1 -> NSwag.Core 13.17.1 -> Newtonsoft.Json 9.0.1 -> System.Xml.ReaderWriter 4.0.11 -> System.IO.FileSystem 4.0.1 -> runtime.win.System.IO.FileSystem 4.3.0 -> System.Collections (>= 4.3.0) 
 MiSystems.GeoServices -> NSwag.AspNetCore 13.17.1 -> NSwag.Core 13.17.1 -> Newtonsoft.Json 9.0.1 -> System.Collections (>= 4.0.11)	MiSystems.GeoServices	C:\Work\MiSystems\MiSystems.GeoServices\MiSystems.GeoServices.csproj	
```

**Comment by trejjam** [comment]
Hello @r-work and thanks.
It should be solvable in your code by declaring a dependency on `System.Xml.ReaderWriter=4.3.0` and `System.Collection=4.3.0`, or better on a higher `Newtonsoft.Json`. Which you should do anyway, there are some security issues.

At the same time, I bumped the minimal version in this library to 10.0.1, which is still vulnerable to the security issues mentioned above, but it should resolve without help.

**Comment by prom3theu5** [comment]
Aye that’s right. Anything below 13.0.1 synk or another Analyzers will raise CWE-755

**Comment by r-work** [comment]
Hi @trejjam, sorry had to delete my previous comment as I realized it contained internal info.

Our project does not reference `Newtonsoft.Json` as we've migrated to `System.Text.Json`.

So the issue might just be NSwag referencing an old `Newtonsoft.Json` version.

**Comment by trejjam** [comment]
Original message:

> This fails if your project is net 7 and have <PlatformTarget>x64</PlatformTarget> in your csproj with the following message:
> 
>  Detected package downgrade: System.Collections from 4.3.0 to 4.0.11. Reference the package directly from the project to select a different version. 
>  `A.Project -> NSwag.AspNetCore 13.17.1 -> NSwag.Core 13.17.1 -> Newtonsoft.Json 9.0.1 -> System.Xml.ReaderWriter 4.0.11 ->  System.IO.FileSystem 4.0.1 -> runtime.win.System.IO.FileSystem 4.3.0 -> System.Collections (>= 4.3.0)`
>  `A.Project -> NSwag.AspNetCore 13.17.1 -> NSwag.Core 13.17.1 -> Newtonsoft.Json 9.0.1 -> System.Collections (>= 4.0.11)	A.Project	C:\A.Project\A.Project.csproj`

> Our project does not reference Newtonsoft.Json as we've migrated to System.Text.Json.

Since you are using NSwag you are using Newtonsoft.Json, so you should upgrade it. Lower-bound dependencies are kinda a habit here :) (They permit you to upgrade when you want, instead of forcing you)

**Comment by r-work** [comment]
I meant it's not referencing it directly.
Got it, thanks.

**Comment by ilias-hamdaoui** [comment]
Thank you @RicoSuter  for the release !

**Comment by RicoSuter** [comment]
v13.18.0 is now released with .NET 7... please report if it works as expected

**Comment by NoahStolk** [comment]
I'm able to upgrade to .NET 7 now using NSwag 13.18.0. All seems to work as expected. Thank you for the release. 🚀

**Comment by RicoSuter** [comment]
@trejjam can you tell me why the BodyParametersTests are invalid in .NET 7 (Skip = "Wrong in .Net 7")?

**Comment by trejjam** [comment]
Hi @RicoSuter,

The test says `When_primitive_body_parameter_has_default_value_then_it_is_optional` but it's tested to be required in .Net6. In .Net 7 `IsRequired` result is false for `BodyParameters_RequiredPrimitiveWithDefault` (as it should be, based on the test description)

**Comment by trejjam** [comment]
"Correct" solution, in my opinion, is probably to remove `*before_net7` with conditional `Skip` definition `!NET7_0_OR_GREATER` and then fix the parser for .Net6. The question is if it's worthy/possible.

**Comment by RicoSuter** [comment]
Ok, will think about it. Now working on the bigger v14 release itself...

**Pull request merged**